below/model/src/tc_model.rs

```diff
@@ -0,0 +1,267 @@
+use std::time::Duration;
+
+use serde::{Deserialize, Serialize};
+
+use crate::{Field, FieldId, Nameable, Queriable};
+use tc::{QDisc, Tc, XStats};
+
+/// rate! macro calculates the rate of a field for given sample and last objects.
+/// It basically calls count_per_sec! macro after extracting the field from the objects.
+macro_rules! rate {
+    ($field:ident, $sample:ident, $last:ident, $target_type:ty) => {{
+        $last.and_then(|(last, d)| {
+            let s = $sample.$field;
+            let l = last.$field;
+            count_per_sec!(l, s, d, $target_type)
+        })
+    }};
+}
+
+#[derive(
+    Clone,
+    Debug,
+    Default,
+    PartialEq,
+    Serialize,
+    Deserialize,
+    below_derive::Queriable
+)]
+pub struct TcModel {
+    #[queriable(subquery)]
+    pub tc: Vec<SingleTcModel>,
```

Comment:
`TcModel` is queriable on index. Is there some field in `SingleTcModel` that uniquely identifies the `SingleTcModel` (e.g. `interface?`). If so we should pull it out and make this a `BTreeMap<String, SingleTcModel>`, so that it's easier to query.

Reply:
It is not straightforward because there's no clear identifier for qdiscs, also an interface could have multiple qdiscs.
From this [website](http://borg.uu3.net/traffic_shaping/classes.html#:~:text=qdiscs%20are%20always%20referred%20to%20using%20a%20combination%20of%20a%20major%20node%20number%20and%20a%20minor%20node%20number.%20For%20any%20given%20qdisc%2C%20the%20major%20node%20number%20has%20to%20be%20unique%20for%20the%20root%20hook%20for%20a%20given%20Ethernet%20interface.):
> qdiscs are always referred to using a combination of a major node number and a minor node number. For any given qdisc, the major node number has to be unique for the root hook for a given Ethernet interface.

I believe we could use a combination of interface, handle and parent to uniquely identify a qdisc. We could add a `key` field and assign to it the hashed or concatenated combination of the 3 fields.

Reply:
It seems to me like it should be keyed uniquely identified by major/minor which are actually included in `TcMessage`. I happen to already have a bit of code handy (we use it for identifying GPUs in the internal version).

You just need to change this to `BTreeMap<MajMin, SingleTcModel>`.

```
use serde_with::DeserializeFromStr;
use serde_with::SerializeDisplay;

// A Queriable MajMin
#[derive(
    Default,
    Clone,
    PartialEq,
    PartialOrd,
    Ord,
    Eq,
    Debug,
    DeserializeFromStr,
    SerializeDisplay,
    below_derive::Queriable
)]
pub struct MajMin {
    pub minor_id: u64,
    pub major_id: u64,
}

impl std::fmt::Display for MajMin {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}:{}", self.major_id, self.minor_id)
    }
}

impl std::str::FromStr for MajMin {
    type Err = anyhow::Error;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if let Some(colon_idx) = s.find(':') {
            Ok(Self {
                major_id: (&s[..colon_idx]).parse::<u64>()?,
                minor_id: (&s[colon_idx + 1..]).parse::<u64>()?,
            })
        } else {
            Err(anyhow!("No colon in 'maj:min'"))
        }
    }
}
```